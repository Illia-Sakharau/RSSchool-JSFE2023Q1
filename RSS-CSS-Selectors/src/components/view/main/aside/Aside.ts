import './aside.scss';
import createElement from '../../../../utils/createElement';
import levels from '../../../../data/levels';

export default class Aside {
  public draw(): HTMLElement {
    const asideEl: HTMLElement = createElement({ tag: 'aside' });
    const title: HTMLElement = createElement({ tag: 'h6', content: 'Levels' });
    const levelWrapperel: HTMLElement = createElement({ tag: 'div', classes: ['levels-wrapper'] });
    const buttonEl: HTMLElement = createElement({ tag: 'button', classes: ['button'], content: 'Reset progress' });

    const activeLevel = localStorage.getItem(`ily-active`) || 0;

    asideEl.appendChild(title);

    levels.forEach((lev, ind) => {
      const classes = ['level-btn'];
      const state = localStorage.getItem(`ily-level${ind}`);
      if (state) {
        classes.push(state);
      }
      if (+activeLevel === ind) {
        classes.push('active');
      }
      const levelEl = createElement({ tag: 'button', classes: classes, content: `Level ${ind + 1}` });
      levelWrapperel.appendChild(levelEl);

      levelEl.addEventListener('click', () => console.log(`Level ${ind + 1}`));
    });
    asideEl.appendChild(levelWrapperel);

    buttonEl.addEventListener('click', () => console.log(`Reset progress`));
    asideEl.appendChild(buttonEl);

    // open/close aside meny
    document.addEventListener('toggleMenu', () => {
      console.log(asideEl);
      asideEl.classList.toggle('visible');
    });

    return asideEl;
  }
}
