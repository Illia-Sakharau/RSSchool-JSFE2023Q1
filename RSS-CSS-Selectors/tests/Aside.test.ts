import Aside from '../src/components/view/main/aside/Aside';

jest.mock('../src/data/levels', () => {
  return [{}, {}, {}];
});
jest.mock('../src/data/constants', () => {
  return {
    CURRENT_LEVEL: {
      getLevel: () => 1,
    },
    LEVELS_STATES: ['completed_help', null, 'completed'],
  };
});
jest.mock('../src/components/view/main/aside/aside.scss', () => ({}));

const aside = new Aside();
const asideElement = aside.draw();

afterAll(() => {
  jest.restoreAllMocks();
});

describe('Test drow aside', () => {
  test('should return correct answer', () => {
    expect(asideElement).toBeDefined();
  });
  test('should create 3 level buttons', () => {
    expect(asideElement.querySelectorAll('.level-btn')).toHaveLength(3);
  });
  test('should create 1 level button as completed with helper', () => {
    expect(asideElement.querySelector('.completed_help')?.textContent).toEqual('Level 1');
  });
  test('should create 2 level button as active', () => {
    expect(asideElement.querySelector('.active')?.textContent).toEqual('Level 2');
  });
  test('should create 3 level button as completed', () => {
    expect(asideElement.querySelector('.completed')?.textContent).toEqual('Level 3');
  });
  test('should return aside', () => {
    expect(asideElement).toMatchSnapshot();
  });
});
