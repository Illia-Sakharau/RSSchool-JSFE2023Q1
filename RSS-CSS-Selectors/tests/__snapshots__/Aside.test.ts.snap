// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test drow aside should return aside 1`] = `
<aside>
  <h6>
    Levels
  </h6>
  <div
    class="levels-wrapper"
  >
    <button
      class="level-btn completed_help"
      data-level="0"
    >
      Level 1
    </button>
    <button
      class="level-btn active"
      data-level="1"
    >
      Level 2
    </button>
    <button
      class="level-btn completed"
      data-level="2"
    >
      Level 3
    </button>
  </div>
  <button
    class="button"
  >
    Reset progress
  </button>
</aside>
`;
